### PR TITLE
Implement context-specific tags on #[derive(Sequence)]

### LINKED ISSUE #322: `der` crate fails to emit implicit context-specific tags on `SET`
There is a bug with the generated DER somewhere in the `der` crate. I discovered this while writing a `pkcs10` crate. This causes the encoding tests to fail. Specifically, it fails to correctly emit the `IMPLICIT` tag modification of the `SET` in `CertReqInfo::attributes` (see tagged PR). Decoding works correctly.

The fix is probably simple, but I don't know enough about the structure of the `der` crate to find it easily. And since I'm not sure you actually want this PR, I didn't spend a lot of time on it. I'm willing to produce a patch if you point me in the right direction.

The following code triggers the bug:

```
#[derive(Clone, Debug, PartialEq, Eq, Sequence)]
pub struct Foo {
    #[asn1(context_specific = "0", tag_mode = "IMPLICIT")]
    pub bar: SetOfVec<Baz>,
}
```

Decoding of the above code works correctly. But encoding causes `SET` to be emitted without the context-specific `0` tag.

### PR DESCRIPTION
**Pull request by bstrie**
This also adds a context-specific tag test to the code for #[derive(Choice)], and cleans up the tests in these files.

If you'd like I would be happy to add more tests to cover the interactions between the various options, but I admit that I'm not an ASN.1 expert so I would need some guidance as to how these options interact so that I could formulate the correct desugaring.

Closes #322.

### CODE REVIEW

**Comment by npmccallum** [comment]
@bstrie I think you mean context specific, not context sensitive.

**Comment by tarcieri** [comment]
> I think you mean context specific, not context sensitive

Thanks to the Chomsky Hierarchy I have made this mistake several times myself 😅

File: der/derive/src/choice/variant.rs [hunk 1]
```diff
@@ -97,25 +98,91 @@ mod tests {
...
         assert_eq!(
             variant.to_decode_tokens().to_string(),
-            ":: der :: Tag :: Utf8String => Ok (Self :: ExampleVariant (decoder . decode () ? . try_into () ?)) ,"
+            quote! {
+                ::der::Tag::Utf8String => Ok(Self::ExampleVariant(
+                    decoder.decode()?
+                    .try_into()?
+                )),
+            }
+            .to_string()
```
Comment (tarcieri) [inline der/derive/src/choice/variant.rs:109]: Oh wow this is the simple and obvious-in-hindsight solution. Nice

**Comment by npmccallum** [comment]
@bstrie I think you need to fix the commit message wording too.

**Comment by npmccallum** [comment]
@bstrie Also, one of the tests is failing.

**Comment by tarcieri** [comment]
One of the errors is on `IssuingDistributionPointsExample`:

```rust
    impl ::der::Sequence<'_> for IssuingDistributionPointExample {
        fn fields<F, T>(&self, f: F) -> ::der::Result<T>
        where
            F: FnOnce(&[&dyn der::Encodable]) -> ::der::Result<T>,
        {
            ()
        }
    }
```

...not sure what's happening there but the empty tuple is definitely not the right result.

Looks like something similar on the other error:

```
    impl<'a> ::der::Sequence<'a> for OneAsymmetricKey<'a> {
        fn fields<F, T>(&self, f: F) -> ::der::Result<T>
        where
            F: FnOnce(&[&dyn der::Encodable]) -> ::der::Result<T>,
        {
            ()
        }
    }
```

Perhaps the unparseable tokens are being lost?

```
error: proc-macro derive produced unparseable tokens
```

File: der/derive/src/sequence/field.rs [hunk 2]
```diff
@@ -188,13 +192,29 @@ impl LowerFieldEncoder {
...
+        let encoder = &self.encoder;
+        let number_tokens = tag_number.to_tokens();
+        let mode_tokens = tag_mode.to_tokens();
+
+        self.encoder = quote! {
+            &::der::asn1::ContextSpecific {
+                tag_number: #number_tokens,
+                tag_mode: #mode_tokens,
+                value: Clone::clone(#encoder),
+            },
```
Comment (tarcieri) [inline der/derive/src/sequence/field.rs:207]: I think the `,` here needs to be removed:

```suggestion
            }
```

The proc macro assembles these together using `quote`:

```rust
f(&[
    #(#encode_body),*
])
```

File: der/derive/src/sequence/field.rs [hunk 3]
```diff
@@ -188,13 +192,29 @@ impl LowerFieldEncoder {
...
     }
+
+    /// Make this field context-specific.
+    fn apply_context_specific(&mut self, tag_number: &TagNumber, tag_mode: &TagMode) {
+        let encoder = &self.encoder;
+        let number_tokens = tag_number.to_tokens();
+        let mode_tokens = tag_mode.to_tokens();
+
+        self.encoder = quote! {
+            &::der::asn1::ContextSpecific {
```
Comment (tarcieri) [inline der/derive/src/sequence/field.rs:203]: The `ContextSpecificRef` type should make the `Clone` unnecessary:

https://docs.rs/der/latest/der/asn1/struct.ContextSpecificRef.html

**Comment by bstrie** [comment]
Code updated to pass all the tests (sorry about not doing so the first time, I forgot to test with `--all-features`). The tests that were failing were already using context-specific tags (which I guess were just being ignored before now) in conjunction with various other field attributes. Taking into account all the supported combinations of the currently-implemented attributes was a bit tricky and required some subtle changes to make sure all the references lined up correctly, but in the end I think I managed to find a nice arrangement that doesn't involve adding state to the Lowerer or else compromising the unit-testability of the Lowerer methods. Once again I tested this with https://github.com/RustCrypto/formats/pull/323 to confirm that it addresses https://github.com/RustCrypto/formats/issues/322 .

File: der/tests/derive.rs [hunk 4]
```diff
@@ -353,6 +353,28 @@ mod sequence {
     const ALGORITHM_IDENTIFIER_DER: &[u8] =
         &hex!("30 13 06 07 2a 86 48 ce 3d 02 01 06 08 2a 86 48 ce 3d 03 01 07");
 
+    #[derive(Sequence)]
+    #[asn1(tag_mode = "IMPLICIT")]
+    pub struct TypeCheckExpandedSequenceFieldAttributeCombinations<'a> {
```
Comment (tarcieri) [inline der/tests/derive.rs:358]: Oh my 😅

**Review by tarcieri — APPROVED** [review]

File: der/tests/derive.rs (hunk 4 above)
Comment (bstrie) [reply der/tests/derive.rs:358]: We call that "self-documenting code". :P

**Pull request merged**